auth/README.md

```diff
@@ -336,6 +354,78 @@ see the Firebase Auth documentation to
+Eventually, they switch devices and you automatically sign them in anonymously to give your users a
+frictionless UX by letting users interact with your app without being forced to log in immediately.
+Now, your anonymously signed in user quickly enters some data into your app before signing into
+their existing account so as not to forget why they came to your app in the first place. This UX is
+great for users, but problematic for us developers: your user now has two different ids that match
+the same person! The above example isn't just limited to anonymous accounts; **anytime a user has
+two existing accounts, a user collision will occur and you will have to manually merge those
+accounts together**. To help you do this, FirebaseUI provides a service to guide you through to
+process of loading and the transferring the user's data.
+
+The following is an example of how you would move data from the user's previous uid to their new one
+using the Firebase Realtime database, thus merging the two accounts.
+The example assumes you are using data structured similarly to the
+[sample](https://github.com/firebase/FirebaseUI-Android/blob/master/app/src/main/java/com/firebase/uidemo/database/ChatHolder.java):
+
+```java
+public class MyManualMergeService extends ManualMergeService {
+    private Iterable<DataSnapshot> mChatKeys;
+
+    @Override
+    public Task<Void> onLoadData() {
+        final TaskCompletionSource<Void> loadTask = new TaskCompletionSource<>();
+        FirebaseDatabase.getInstance()
+                .getReference()
+                .child("chatIndices")
+                .child(FirebaseAuth.getInstance().getCurrentUser().getUid())
+                .addListenerForSingleValueEvent(new ValueEventListener() {
+                    @Override
+                    public void onDataChange(DataSnapshot snapshot) {
+                        mChatKeys = snapshot.getChildren();
+                        loadTask.setResult(null);
+                    }
+
+                    @Override
+                    public void onCancelled(DatabaseError error) {
+                        FirebaseCrash.report(error.toException());
+                    }
+                });
+        return loadTask.getTask();
+    }
+
+    @Override
+    public Task<Void> onTransferData(IdpResponse response) {
+        String uid = FirebaseAuth.getInstance().getCurrentUser().getUid();
+        DatabaseReference chatIndices = FirebaseDatabase.getInstance()
+                .getReference()
+                .child("chatIndices")
+                .child(uid);
+        for (DataSnapshot snapshot : mChatKeys) {
+            chatIndices.child(snapshot.getKey()).setValue(true);
+            DatabaseReference chat = FirebaseDatabase.getInstance()
+                    .getReference()
+                    .child("chats")
+                    .child(snapshot.getKey());
+            chat.child("uid").setValue(uid);
```

Comment:
Hmm... If I understand it correctly, your database example isn't using Rules to enforce user based security?
https://firebase.google.com/docs/database/security/user-security

If really enforcing user based security, there are only two options to migrate the data:
1. Back up all data in memory and then copy to the new (non-anon) user session
2. Having two user session (two FirebaseDatabase instances w/ different users) side by side and do the copying.

Reply:
See https://github.com/firebase/FirebaseUI-Android/issues/123#issuecomment-319178534, I haven't updated the docs in a bit.

Reply:
Thanks @SUPERCILEX. I see, it works for your data structure. But I'm afraid this "anon-user session only" model might not be sufficient for other database structures.

But I do like your assisted merging idea! Thanks for helping all developers here! 

Let's continue the discussion (offline) how to make this feature work. At high level, I think we can allow developers choose from below three options:
a) Replace: Throw away anonymous user (same as today)
b) Link if possible or fail (and give back Credential for developers to implement their own merging logic) 
c) Link or replace with assisted merging

c) might be difficult to implement if we need to make it work for all database structures.

Reply:
@isaisachen Sorry for the late reply! 😊

K, I just realized you're a FirebaseUI iOS dev—thanks for getting this going! Since I thought you were a user, I didn't really explain much so let me try and fix that.

The link I sent you is actually for the extreme case where the user is playing around with three different accounts, which I've documented in much more detail [here](https://github.com/SUPERCILEX/FirebaseUI-Android/blob/cfe884fa8b6ce516d9b2310cf9a09cfec33b6c24/auth/src/main/java/com/firebase/ui/auth/data/model/CyclicAccountLinkingException.java). For normal use cases, I do perform automatic linking using the Firebase APIs so the original user account isn't thrown away. To address another point, this merging isn't limited to anonymous accounts—it works for any merging of two or more accounts.

As for your database concerns, I don't see how I could make it any more flexible—I'm literally letting the developer execute any piece of code they so desire during that no man's land sign-in period! 😄 However, I really like your idea of giving the developer the old credential too so we can cover all the possible cases (I already give them the new one).

@isaisachen Does that answer some of your questions? How can I help you build this feature?

Reply:
Hi @SUPERCILEX. Thanks for the reply. I'm not FirebaseUI iOS dev :) But @samtstern will reach out.